File: implementacion_java/src/algoritmos/WinogradOriginal.java
package algoritmos;

import utilidades.AlgoritmoMultiplicacion;
import utilidades.Convertidor;

import java.math.BigInteger;

public class WinogradOriginal implements AlgoritmoMultiplicacion {

    Convertidor convertidor = new Convertidor();

    @Override
    public void multiplicar(BigInteger[][] matriz1, BigInteger[][] matriz2) {
        // Convertir de BigInteger a int
        int[][] doubleMatriz1 = convertidor.convertBigIntegerArrayToIntArray(matriz1);
        int[][] doubleMatriz2 = convertidor.convertBigIntegerArrayToIntArray(matriz2);
        int[][] doubleResultado = multiplicarWinogradOriginal(doubleMatriz1, doubleMatriz2);

        // Convertir el resultado double a BigINteger
        convertidor.convertIntArrayToBigInteger(doubleResultado);
    }

    public int[][] multiplicarWinogradOriginal(int[][] matrizA, int[][] matrizB) {
        int N = matrizA.length; // Número de filas de A
        int P = matrizA[0].length; // Número de columnas de A (y filas de B)
        int M = matrizB[0].length; // Número de columnas de B
        int[][] matrizC = new int[N][M]; // Matriz de resultado
        int[] y = new int[M]; // Vector de factores de fila
        int[] z = new int[N]; // Vector de factores de columna

        int aux; // Variable auxiliar para acumulaciones
        int upsilon = P % 2; // Determina si P es par o impar
        int gamma = P - upsilon; // Ajusta gamma para manejar pares

        int i, j, k;

        // Precalcular los factores de filas
        for (i = 0; i < M; i++) {
            aux = 0;
            for (j = 0; j < gamma; j += 2) {
                aux += matrizA[i][j] * matrizA[i][j + 1]; // Sumar productos de pares de elementos en la fila
            }
            y[i] = aux;
        }

        // Precalcular los factores de columnas
        for (i = 0; i < N; i++) {
            aux = 0;
            for (j = 0; j < gamma; j += 2) {
                aux += matrizB[j][i] * matrizB[j + 1][i]; // Sumar productos de pares de elementos en la columna
            }
            z[i] = aux;
        }

        if (upsilon == 1) { // Si P es impar
            /*
             * Si P es impar, falta el término matrizA[i][P] * matrizB[P][k] en todas las sumas auxiliares.
             */
            int PP = P - 1; // Última posición válida de los índices
            for (i = 0; i < M; i++) {
                for (k = 0; k < N; k++) {
                    aux = 0;
                    for (j = 0; j < gamma; j += 2) {
                        aux += (matrizA[i][j] + matrizB[j + 1][k]) * (matrizA[i][j + 1] + matrizB[j][k]); // Producto ajustado
                    }
                    matrizC[i][k] = aux - y[i] - z[k] + matrizA[i][PP] * matrizB[PP][k]; // Ajustar el término faltante
                }
            }
        } else { // Si P es par
            /*
             * Si P es par, el resultado se puede calcular directamente con las sumas auxiliares.
             */
            for (i = 0; i < M; i++) {
                for (k = 0; k < N; k++) {
                    aux = 0;
                    for (j = 0; j < gamma; j += 2) {
                        aux += (matrizA[i][j] + matrizB[j + 1][k]) * (matrizA[i][j + 1] + matrizB[j][k]); // Producto ajustado
                    }
                    matrizC[i][k] = aux - y[i] - z[k]; // Usar las sumas auxiliares
                }
            }
        }

        // Devolver la matriz resultado
        return matrizC;
    }
}

File: implementacion_kotlin/src/multiplicacion_matrices/WinogradOriginal.kt
package multiplicacion_matrices

import utilidades.AlgoritmoMultiplicacion
import utilidades.Convertidor
import java.math.BigInteger

class WinogradOriginal : AlgoritmoMultiplicacion {
    var convertidor: Convertidor = Convertidor()

    override fun multiplicar(matriz1: Array<Array<BigInteger>>, matriz2: Array<Array<BigInteger>>) {
        // Convertir de BigInteger a int
        val doubleMatriz1 = convertidor.convertBigIntegerArrayToIntArray(matriz1)
        val doubleMatriz2 = convertidor.convertBigIntegerArrayToIntArray(matriz2)
        val doubleResultado = multiplicarWinogradOriginal(doubleMatriz1, doubleMatriz2)

        // Convertir el resultado double a BigINteger
        convertidor.convertIntArrayToBigInteger(doubleResultado)
    }

    /**
     * Permite multiplicar dos matrices mediante el método WinogradOriginal
     *
     * @param matrizA primera matriz
     * @param matrizB segunda matriz
     * @return matrizC matriz resultado de la multiplicación entre A y B
     */
    fun multiplicarWinogradOriginal(matrizA: Array<IntArray>, matrizB: Array<IntArray>): Array<IntArray> {
        val N = matrizA.size
        val P = matrizA[0].size
        val M = matrizB[0].size
        val matrizC = Array(N) { IntArray(M) }
        val y = IntArray(M)
        val z = IntArray(N)

        var aux: Int
        val upsilon = P % 2
        val gamma = P - upsilon

        var i: Int
        var j: Int
        var k: Int

        i = 0
        while (i < M) {
            aux = 0
            j = 0
            while (j < gamma) {
                aux += matrizA[i][j] * matrizA[i][j + 1]
                j += 2
            }
            y[i] = aux
            i++
        }
        i = 0
        while (i < N) {
            aux = 0
            j = 0
            while (j < gamma) {
                aux += matrizB[j][i] * matrizB[j + 1][i]
                j += 2
            }
            z[i] = aux
            i++
        }
        if (upsilon == 1) {
            /*
			 * P is odd The value matrizA[i][P]*matrizB[P][k] is missing in all auxiliary
			 * sums.
			 */
            val PP = P - 1
            i = 0
            while (i < M) {
                k = 0
                while (k < N) {
                    aux = 0
                    j = 0
                    while (j < gamma) {
                        aux += (matrizA[i][j] + matrizB[j + 1][k]) * (matrizA[i][j + 1] + matrizB[j][k])
                        j += 2
                    }
                    matrizC[i][k] = aux - y[i] - z[k] + matrizA[i][PP] * matrizB[PP][k]
                    k++
                }
                i++
            }
        } else {
            /*
			 * P is even The result can be computed with the auxiliary sums.
			 */
            i = 0
            while (i < M) {
                k = 0
                while (k < N) {
                    aux = 0
                    j = 0
                    while (j < gamma) {
                        aux += (matrizA[i][j] + matrizB[j + 1][k]) * (matrizA[i][j + 1] + matrizB[j][k])
                        j += 2
                    }
                    matrizC[i][k] = aux - y[i] - z[k]
                    k++
                }
                i++
            }
        }
        return matrizC
    }
}
